### ISSUE TITLE
Add support for tfenv

### ISSUE THREAD

**Issue by bobh66** [upbound/provider-terraform#84]
### What problem are you facing?
This is for discussion of the changes proposed in #82 by @bdwyertech

The changes are significant enough that I'd like to have them documented/discussed here as well as in the PR.  

As I understand it there is a need to support multiple versions of terraform within the same provider-terraform pod/container and to be able to select the desired version of terraform for each individual Workspace.

### How could Official Terraform Provider help solve your problem?
tfenv supplies a set of wrapper shell scripts that use environment variables to select the desired/required version of terraform.

The proposed changes install tfenv from git using git clone and link the tfenv shell scripts into /usr/local/bin, which includes a shell version of the "terraform" executable.  tfenv is then used to download the specific version of terraform for the provider into /opt/tfenv/versions where it can be executed by the terraform wrapper script.

I'd like to understand how the user would specify the version of terraform that they want to use and what that means functionally within the provider.  Does tfenv automatically download the desired version of terraform the first time is it requested?  Does that happen automatically when the terraform script is invoked with the specified env variable?

I realize these changes aren't major from an implementation standpoint, but we (the maintainers) have to maintain them and so it's important to have everything clearly documented and understood so there aren't any surprises later on.

Also if the feature is going to be useful to others who may not be familiar with tfenv it is important to have it fully documented so it's clear how to take advantage of the functionality.

**Comment by bdwyertech** [comment]
> I'd like to understand how the user would specify the version of terraform that they want to use and what that means functionally within the provider. Does tfenv automatically download the desired version of terraform the first time is it requested? Does that happen automatically when the terraform script is invoked with the specified env variable?

Yes.  When invoked, tfenv transparently pulls the version specified.  If one chooses to do so, they can even pass an env var to configure the source.  This is amenable to our situation running in an air-gap.  Can be pointed to Nexus/Artifactory/etc.

> I realize these changes aren't major from an implementation standpoint, but we (the maintainers) have to maintain them and so it's important to have everything clearly documented and understood so there aren't any surprises later on.
> 
We’re either going to contribute back or run our own fork.  Would obviously prefer to participate here and avoid the latter.  Would be open to implementing similar logic in Go but I’m a believer in KISS and not reinventing the wheel. Tfenv is well known and understood by the Terraform community.

**Comment by jeanduplessis** [comment]
At this point in time, we will not allow overriding the Terraform binary with the possibility of one released after 1.5.x as they are released under the BSL license. This provider would violate the license terms, and we do not want to open the possibility for that even if the action was taken by a user of the provider.
